x/concentrated-liquidity/swaps_test.go

```diff
@@ -1361,3 +1242,85 @@ func (s *KeeperTestSuite) TestSwapExactAmountOut() {
+
+			// Create default CL pool
+			pool := s.PrepareConcentratedPool()
+
+			// add default position
+			_, _, _, err := s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, pool.GetId(), s.TestAccs[0], DefaultAmt0, DefaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), DefaultLowerTick, DefaultUpperTick)
+			s.Require().NoError(err)
+
+			// add second position depending on the test
+			if !test.secondPositionLowerPrice.IsNil() {
+				newLowerTick, err := math.PriceToTick(test.secondPositionLowerPrice, DefaultExponentAtPriceOne) // 315000
+				s.Require().NoError(err)
+				newUpperTick, err := math.PriceToTick(test.secondPositionUpperPrice, DefaultExponentAtPriceOne) // 322500
+				s.Require().NoError(err)
+
+				_, _, _, err = s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, pool.GetId(), s.TestAccs[1], DefaultAmt0, DefaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), newLowerTick.Int64(), newUpperTick.Int64())
+				s.Require().NoError(err)
+			}
+
+			// mark pool state and user balance before swap
+			poolBefore, err := s.App.ConcentratedLiquidityKeeper.GetPool(s.Ctx, pool.GetId())
+			s.Require().NoError(err)
+			userBalanceBeforeSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, s.TestAccs[0])
+			poolBalanceBeforeSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, poolBefore.GetAddress())
+
+			// system under test
+			firstTokenIn, firstTokenOut, _, _, _, err := s.App.ConcentratedLiquidityKeeper.SwapOutAmtGivenIn(
+				s.Ctx,
+				test.tokenIn, test.tokenOutDenom,
+				DefaultZeroSwapFee, test.priceLimit, pool.GetId())
+
+			secondTokenIn, secondTokenOut, _, _, _, err := s.App.ConcentratedLiquidityKeeper.SwapOutAmtGivenIn(
+				s.Ctx,
+				firstTokenOut, firstTokenIn.Denom,
+				DefaultZeroSwapFee, sdk.ZeroDec(), pool.GetId(),
+			)
+			s.Require().NoError(err)
+
+			// INVARIANTS
+
+			// 1. assure we get the same tokens after swapping back and forth
+
+			// allow 0.01% of margin of error
+			firstTokenInMargin := firstTokenIn.Amount.Mul(sdk.NewInt(1)).Quo(sdk.NewInt(10000))
+			tokenInDiff := firstTokenIn.Amount.Sub(secondTokenOut.Amount).Abs()
+
+			s.Require().True(firstTokenInMargin.GTE(tokenInDiff))
+			s.Require().Equal(firstTokenOut, secondTokenIn)
+
+			// 2. assure that pool state came back to original state
+			poolAfter, err := s.App.ConcentratedLiquidityKeeper.GetPool(s.Ctx, pool.GetId())
+			s.Require().NoError(err)
+
+			s.Require().Equal(poolBefore.GetTotalShares(), poolAfter.GetTotalShares())
+			s.Require().Equal(poolBefore.GetTotalPoolLiquidity(s.Ctx), poolAfter.GetTotalPoolLiquidity(s.Ctx))
+
+			oldSpotPrice, err := poolBefore.SpotPrice(s.Ctx, pool.GetToken0(), pool.GetToken1())
+			s.Require().NoError(err)
+			newSpotPrice, err := poolAfter.SpotPrice(s.Ctx, pool.GetToken0(), pool.GetToken1())
+			s.Require().NoError(err)
+			oldSpotPriceMarginUp := oldSpotPrice.Mul(sdk.MustNewDecFromStr("0.001"))
+			spotPriceDiff := newSpotPrice.Sub(oldSpotPrice).Abs()
+			s.Require().True(oldSpotPriceMarginUp.GT(spotPriceDiff))
```

Comment:
Same question about using osmoutils for this (maybe it does the same thing, but imo its cleaner / more obvious / less prone to miscalculating.

Reply:
Ditto here, done as well